### PR TITLE
Implement hover and (stubbed) codeAction on stale state

### PR DESCRIPTION
**Pull request by aprocter**
### Motivation
This implements a (beta, or let's even call it alpha) feature that allows `hover` queries to proceed even when slow-path typechecking is in progress. You have to opt into the feature with `--enable-experimental-lsp-stale-state`.

### Test plan
See included automated tests.

### CODE REVIEW

File: main/lsp/LSPTypechecker.cc
Comment (froydnj) [inline main/lsp/LSPTypechecker.cc:696]: This is not safe, right?  We don't want `undoState.getEvictedGs` to be modified, yet we're passing it to a function that's going to modify it.  (`getResolved` has the same issue.)

Comment (aprocter) [reply main/lsp/LSPTypechecker.cc:696]: Since we're pulling in the evicted file, which was already indexed relative to the evicted gs, I think we shouldn't actually be introducing any new symbols and therefore this should be de facto safe. (Not 100% confident though, so I should also note that this is hidden behind an opt-in CLI flag.)

File: main/lsp/LSPTypechecker.h
Comment (jez) [inline main/lsp/LSPTypechecker.h:208]: It looks like there's an asymmetry between `LSPStaleTypechecker` and `LSPTypecheckerDelegate` here.

Why doesn't `LSPTypecheckerDelegate` take an `LSPConfiguration` or a `WorkerPool`?

Comment (aprocter) [reply main/lsp/LSPTypechecker.h:208]: The `config` is used inside `LSPStaleTypechecker::query` directly because `LSPStaleTypechecker::query` largely duplicates the impelmentation of `LSPTypechecker::query`, while `LSPTypecheckerDelegate` can just delegate `LSPTypechecker::query`, where `config` is a field on `LSPTypechecker`.

The `WorkerPool` we create here exists for similar reasons: `LSPTypechecker` has a private `WorkerPool`. (Note that here we create an empty, i.e., zero-thread, `WorkerPool`, which actually causes work to run on the indexing thread itself.)

File: main/lsp/LSPTypechecker.cc
Comment (jez) [reply main/lsp/LSPTypechecker.cc:779]: I don't know that we have an issue with `typecheck`, because `typecheck` already works with a const GlobalState. I think that the only problem is with `getResolved`.

Comment (jez) [reply main/lsp/LSPTypechecker.cc:779]: I think that the idea of using `evictedGs` is a clever idea to not need to borrow the up-to-date GlobalState from the typechecker thread, but I think this confirms that we still want to gate landing this on the non-mutating namer and resolver.

File: main/lsp/requests/hover.cc
Comment (jez) [inline main/lsp/requests/hover.cc:43]: Is there a way we can force the implementation of each LSPMethod to declare whether it's ok to be called with a stale state? I'd love to force opt-in, not opt-out.

(If that was already implemented in a previous PR I'd love to be pointed to the implementation.)

File: main/lsp/LSPTypechecker.cc
Comment (aprocter) [reply main/lsp/LSPTypechecker.cc:779]: > I think this confirms that we still want to gate landing this on the non-mutating namer and resolver.

Yeah, I'd defer to your judgment on that but I think it's probably a fair call to make.

File: main/lsp/requests/hover.cc
Comment (aprocter) [reply main/lsp/requests/hover.cc:43]: There's a virtual method on `LSPTask` that defaults to `return false`, and we'll only feed an `LSPStaleTypechecker` 
 if it returns `true`. Currently only `hover` and `codeAction` opt in.

https://github.com/sorbet/sorbet/blob/db69c5c127d136d49f206e846cbb986cc9818f2f/main/lsp/LSPTask.h#L114-L115

File: main/lsp/LSPTypechecker.h
Comment (jez) [inline main/lsp/LSPTypechecker.h:205]:
```suggestion
    // Using an WorkerPool with size 0 for all typechecker operations causes the work to run on the
    // current thread (usually: the indexer thread).
    std::unique_ptr<WorkerPool> emptyWorkers;
```

**Comment by aprocter** [comment]
🚦 We're going to hold off on merging this til non-mutating namer/resolver are in place, so I've slapped a `[WIP]` on the title here.

**Comment by aprocter** [comment]
https://github.com/sorbet/sorbet/pull/5469/commits/b9cd06d5f93f05abdb24401e6fbda97e212a1afb adds a simple test for a hover query in stale mode.

**Comment by aprocter** [comment]
Pulling this out of `[WIP]` because I think it's ready to go modulo namer/resolver updates. Since everything is hidden behind a flag, I think it'd actually be fine to merge this ahead of the namer/resolver updates, but I'm also fine with holding off until after those are ready.

Flipped a coin to pick a reviewer and jez was the lucky winner. :)

**Comment by aprocter** [comment]
I should also point out that this PR can only access indexed files that were saved into the `UndoState` when the slow path started running. A follow-on change (which I'm planning on opening as a stacked PR in a moment) will allow access to the rest of the indexed files. (Synchronization for that becomes pretty tricky, hence the separate PR.)

File: test/lsp/multithreaded_protocol_test_corpus.cc
Comment (jez) [inline test/lsp/multithreaded_protocol_test_corpus.cc:682]: Is this required? This makes this test even slower than it currently is, right? (iiuc, this makes every slow path operation take 3 seconds artificially)

Comment (aprocter) [reply test/lsp/multithreaded_protocol_test_corpus.cc:682]: I believe a pause is the best option we have, but the slowness is honestly kind of nasty. (Let me get some numbers for you, but I believe that `multithreaded_protocol_test` takes substantially more time than on `master`.)

What we might be able to do is make the wait time configurable and set it to something much quicker than 3 seconds.

Comment (aprocter) [reply test/lsp/multithreaded_protocol_test_corpus.cc:682]: Integrating the mechanism implemented in #5539 gives us:

```
master  32.18s
#5539   34.16s (note: adds test of slow path blocking mechanism, which sleeps for 2s)
PR      39.37s
```

Taking a 7-second hit here is still not wonderful, but I'm thinking it's probably acceptable since we're not planning on adding a huge pile of protocol tests anyway.

(Haven't pushed this to this branch yet, but will do so once #5539 hits `master`.)

Comment (aprocter) [reply test/lsp/multithreaded_protocol_test_corpus.cc:682]: Integrating the mechanism implemented in #5539 gives us:

```
master  32.18s
#5539   34.16s (note: adds test of slow path blocking mechanism, which sleeps for 2s)
PR      39.37s
```

Taking a 7-second hit here is still not wonderful, but I'm thinking it's probably acceptable since we're not planning on adding a huge pile of protocol tests anyway.

(Haven't pushed this to this branch yet, but will do so once #5539 hits `master`.)

**Comment by froydnj** [comment]
I think that since we now use non-mutating namer/resolver, this PR is probably ready for a proper review, is that right @aprocter ?

File: main/lsp/LSPTypechecker.cc
Comment (froydnj) [reply main/lsp/LSPTypechecker.cc:792]: The types here are a bit gross, because `typecheck` doesn't really modify the global state (except for bits that munge the `errorQueue`, but that's already marked `mutable` anyway, but `typecheck` is not typed as such.  I'll fix that up someplace else.

**Review by froydnj — APPROVED** [review]
This all looks reasonable; I think there are some type issues around `pipeline::typecheck` and `getEvictedGs`, but we can fix those up after the fact.

**Pull request merged**
